File: cpp/src/0025.cpp
/*
[25. Reverse Nodes in k-Group (Hard)](https://leetcode.com/problems/reverse-nodes-in-k-group/)

@topic: Linked List
@topic: Recursion

Constraints:
  - The number of nodes in the list is `n`.
  - `1 <= k <= n <= 5000`
  - `0 <= Node.val <= 1000`
*/

#include "structs/LinkedList.hpp"

class Solution {
   public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if (k == 1) return head;
        auto fake = ListNode(0, head);
        auto prev = &fake;
        auto slow = head;
        auto fast = head;
        while (fast) {
            int i = -1;
            while (++i < k && fast) fast = fast->next;
            if (i == k) {
                prev->next = reverse(slow, fast);
                prev = slow;
                slow = fast;
            }
        }
        return fake.next;
    }

   private:
    ListNode* reverse(ListNode* head, ListNode* last) {
        auto prev = last;
        while (head != last) {
            auto next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    using namespace LinkedList;

    Solution o;

    vector<tuple<ListNode*, int, ListNode*>> CASES = {
        {of({1, 2, 3, 4}), 2, of({2, 1, 4, 3})},
        {of({1, 2, 3, 4, 5}), 2, of({2, 1, 4, 3, 5})},
        {of({1, 2, 3, 4, 5}), 3, of({3, 2, 1, 4, 5})},
    };

    for (auto& [head, k, excepted] : CASES) {
        assert(eq(o.reverseKGroup(head, k), excepted));
    }

    auto start = system_clock::now();
    for (auto i = 0; i < 100000; ++i) {
        o.reverseKGroup(of({1, 2, 3, 4, 5}), 2);
        o.reverseKGroup(of({1, 2, 3, 4, 5}), 3);
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: java/src/0025.java
import static org.assertj.core.api.Assertions.assertThat;
import static structs.util.LinkedList.of;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import structs.linkedlist.ListNode;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        var fake = new ListNode(0, head);
        var prev = fake;
        var slow = head;
        var fast = head;
        while (fast != null) {
            var step = -1;
            while (++step < k && fast != null) {
                fast = fast.next;
            }
            if (step == k) {
                head = reverse(slow, fast);
                prev.next = head;
                prev = slow;
                slow = fast;
            }
        }
        return fake.next;
    }

    private ListNode reverse(ListNode head, ListNode tail) {
        var prev = tail;
        var curr = head;
        while (curr != tail) {
            var next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    public static void main(String[] args) {
        var o = new Solution();

        assertThat(o.reverseKGroup(of(1, 2, 3, 4, 5), 2)).isEqualTo(of(2, 1, 4, 3, 5));
        assertThat(o.reverseKGroup(of(1, 2, 3, 4, 5), 3)).isEqualTo(of(3, 2, 1, 4, 5));

        var start = Instant.now();
        for (var i = 0; i < 10000; i++) {
            o.reverseKGroup(of(1, 2, 3, 4, 5), 2);
            o.reverseKGroup(of(1, 2, 3, 4, 5), 3);
        }
        System.out.println(ChronoUnit.MICROS.between(start, Instant.now()));
    }
}
